File: seed.js
const { PrismaClient } = require("@prisma/client");
const xlsx = require("xlsx");
const path = require("path");

const prisma = new PrismaClient();
const filePath = path.join(__dirname, "HINDALCO_1D.xlsx");

// Load Excel file
const workbook = xlsx.readFile(filePath);
const sheetName = workbook.SheetNames[0];
const sheetData = xlsx.utils.sheet_to_json(workbook.Sheets[sheetName]);

async function seedDatabase() {
  try {
    for (const row of sheetData) {
      const datetimeValue = new Date(row.datetime);

      if (isNaN(datetimeValue.getTime())) {
        console.error("Skipping invalid datetime:", row.datetime);
        continue;
      }

      // Check if the record already exists
      const existingRecord = await prisma.stockData.findUnique({
        where: { datetime: datetimeValue },
      });

      if (!existingRecord) {
        await prisma.stockData.create({
          data: {
            datetime: datetimeValue,
            open: parseFloat(row.open),
            high: parseFloat(row.high),
            low: parseFloat(row.low),
            close: parseFloat(row.close),
            volume: parseInt(row.volume, 10),
            instrument: row.instrument || "HINDALCO",
          },
        });
      } else {
        console.log(`Skipping duplicate datetime: ${datetimeValue}`);
      }
    }
    console.log("✅ Data successfully seeded into PostgreSQL!");
  } catch (error) {
    console.error("❌ Error seeding data:", error);
  } finally {
    await prisma.$disconnect();
  }
}

seedDatabase();

File: seed.py
import pandas as pd
from prisma import Prisma
from datetime import datetime

# Initialize Prisma Client
db = Prisma()

async def seed_database():
    await db.connect()
    file_path = "HINDALCO_1D.xlsx"

    try:
        # Load Excel File
        df = pd.read_excel(file_path, engine="openpyxl")

        # Convert datetime column
        df["datetime"] = pd.to_datetime(df["datetime"], errors="coerce")

        # Drop rows with invalid datetime
        df = df.dropna(subset=["datetime"])

        # Insert each row
        for _, row in df.iterrows():
            datetime_value = row["datetime"].to_pydatetime()

            # Check if record already exists
            existing_record = await db.stockdata.find_first(where={"datetime": datetime_value})

            if not existing_record:
                await db.stockdata.create(
                    data={
                        "datetime": datetime_value,
                        "open": float(row["open"]),
                        "high": float(row["high"]),
                        "low": float(row["low"]),
                        "close": float(row["close"]),
                        "volume": int(row["volume"]),
                        "instrument": row.get("instrument", "HINDALCO"),
                    }
                )
            else:
                print(f"Skipping duplicate datetime: {datetime_value}")

        print("✅ Data successfully seeded into PostgreSQL!")

    except Exception as e:
        print(f"❌ Error seeding data: {e}")

    finally:
        await db.disconnect()

# Run the function
import asyncio
asyncio.run(seed_database())
